File: GameTest.py
import numpy as np

from Game import Game


# 测试落子有效性
def test_is_valid_move():
    game = Game(board_size=9)
    # 边界外落子，应无效
    assert not game.is_valid_move(-1, 0)
    assert not game.is_valid_move(9, 0)
    assert not game.is_valid_move(0, -1)
    assert not game.is_valid_move(0, 9)
    # 已有棋子的位置落子，应无效
    game.board[0, 0] = 1
    assert not game.is_valid_move(0, 0)
    print("落子有效性测试通过")


def test_is_eye_pair():
    game = Game(board_size=6)
    game.board = np.array([
        [1, 1, 1, 0, 0, 0],
        [1, 0, 1, 0, 0, 0],
        [1, 1, 0, 1, 0, 0],
        [0, 1, 1, 1, 0, 0],
        [0, 0, 0, 0, 0, 1],
        [0, 0, 0, 1, 1, 1]
    ])
    assert game.is_eye_pair(1, 1, 1, game.board)
    assert game.is_eye_pair(2, 2, 1, game.board)

    game = Game(board_size=6)
    game.board = np.array([
        [1, 1, 1, 0, 0, 0],
        [1, 0, 1, 0, 0, 0],
        [1, 1, 0, 1, 0, 0],
        [0, 1, 1, 1, 1, 1],
        [0, 0, 0, 0, 1, 0],
        [0, 0, 0, 1, 0, 1]
    ])
    assert not game.is_eye_pair(4, 5, 1, game.board)
    assert not game.is_eye_pair(5, 4, 1, game.board)

    game = Game(board_size=6)
    game.board = np.array([
        [0, 1, 0, 1, 0, 1],
        [1, 0, 1, 1, 1, 0],
        [1, 1, 1, 0, 0, 0],
        [0, 0, 1, 1, 1, 1],
        [0, 0, 0, 1, 0, 1],
        [0, 0, 0, 1, 1, 1]
    ])
    assert game.is_eye_pair(0, 0, 1, game.board)
    assert game.is_eye_pair(1, 1, 1, game.board)


def test_is_eye():
    game = Game(board_size=6)
    game.board = np.array([
        [1, 1, 0, 0, 0, 0],
        [1, 0, 1, 0, 0, 0],
        [1, 1, 1, 0, 0, 0],
        [0, 0, 1, 1, 1, 0],
        [0, 0, 0, 1, 0, 1],
        [0, 0, 0, 1, 1, 1]
    ])

    assert game.is_eye(1, 1, 1, game.board)
    assert game.is_eye(4, 4, 1, game.board)

    game = Game(board_size=6)
    game.board = np.array([
        [0, 1, 0, 1, 0, 0],
        [1, 1, 1, 1, 0, 0],
        [1, 1, 1, 0, 0, 0],
        [0, 0, 1, 1, 1, 1],
        [0, 0, 0, 1, 0, 1],
        [0, 0, 0, 1, 1, 1]
    ])

    assert game.is_eye(0, 0, 1, game.board)
    assert game.is_eye(0, 2, 1, game.board)
    assert game.is_eye(4, 4, 1, game.board)

    game = Game(board_size=6)
    game.board = np.array([
        [0, 1, 0, 1, 0, 1],
        [1, 0, 0, 1, 1, 0],
        [1, 1, 1, 0, 0, 0],
        [0, 0, 1, 1, 1, 1],
        [0, 0, 0, 1, 0, 1],
        [0, 0, 0, 1, 1, 1]
    ])
    assert not game.is_eye(0, 0, 1, game.board)
    assert not game.is_eye(0, 4, 1, game.board)

    print("test_is_eye测试通过")


# 测试落子操作
def test_place_stone():
    game = Game(board_size=9)
    x, y = 0, 0
    assert game.make_move(x, y)  # 第一次落子应该成功
    assert game.board[x, y] == 1  # 落子后该位置应该是黑棋（当前玩家为1）
    print("落子操作测试通过")


# 测试 pass 操作
def test_pass_move():
    game = Game(board_size=9)
    game.pass_move()
    assert not game.end_game_check()  # 第一次 pass 不应该结束游戏
    game.pass_move()  # 第二次 pass 应该结束游戏
    assert game.calculate_winner() in [0, 1, 2]  # 终局结果应该是 0（平局）、1（黑方胜）或 2（白方胜）
    print("pass 操作测试通过")


# 测试终局计算得分
def test_calculate_scores():
    game = Game(board_size=3)
    # 手动设置棋盘状态以测试得分计算
    game.board = np.array([
        [1, 0, 2],
        [0, 1, 0],
        [2, 0, 1]
    ])
    black_score, white_score = game.calculate_scores()
    # 根据手动设置的棋盘状态计算预期得分
    # 这里只计算棋子数
    expected_black_score = np.sum(game.board == 1) - game.tie_mu
    expected_white_score = np.sum(game.board == 2) + game.tie_mu
    assert black_score == expected_black_score
    assert white_score == expected_white_score
    print("得分计算测试通过")


# 测试落子后吃子
def test_place_stone_for_remove():
    game = Game(board_size=3)
    # 手动设置棋盘状态以测试得分计算
    game.board = np.array([
        [1, 0, 2],
        [0, 1, 0],
        [2, 0, 1]
    ])
    game.refresh_eat_moves()
    assert game.make_move(0, 1)
    game.render()
    assert game.make_move(1, 0)
    game.render()
    assert game.make_move(1, 2)
    game.render()
    assert game.board[0, 2] == 0
    print("测试落子后吃子通过")


def test_place_stone_for_remove2():
    game = Game(board_size=3)
    # 手动设置棋盘状态以测试得分计算
    game.board = np.array([
        [1, 2, 1],
        [2, 0, 2],
        [1, 2, 1]
    ])
    game.refresh_eat_moves()
    game.make_move(1, 1)
    game.render()
    assert game.board[0, 1] == 0
    assert game.board[1, 0] == 0
    assert game.board[1, 2] == 0
    assert game.board[2, 1] == 0
    print("测试落子后吃子2通过")


def test_single_ko_cycle():
    game = Game(board_size=4)
    game.current_player = 2

    game.board = np.array([
        [2, 0, 0, 1],
        [0, 2, 1, 0],
        [0, 2, 1, 0],
        [2, 1, 0, 1]
    ])

    game.render()
    game.make_move(0, 2)
    game.render()
    game.make_move(0, 1)
    game.render()
    assert not game.is_valid_move(0, 2)
    print("测试单劫通过")


def test_make_move():
    data = """
    . x x x x x . x .
    x x x x x x x . x
    x . x . x x x x x
    x x x x x . x o .
    x x x x x x x x x
    . x x x o x o x x
    x x x o o o o o o
    x x o o o o . o o
    x o o o o o o . o
    """
    game = Game(board_size=9)
    game.parse(data)
    print(game.get_all_valid_moves_include_pass())
    assert len(game.get_all_valid_moves_include_pass()) == 1

    game.render()


def test_ko():
    game = Game(board_size=4)
    game.make_move(0, 0)
    game.render()
    game.make_move(0, 1)
    game.render()
    game = Game(board_size=6)
    game.current_player = 2

    game.board = np.array([
        [2, 0, 0, 1, 0, 0],
        [0, 2, 1, 0, 0, 2],
        [0, 0, 0, 0, 2, 0],
        [0, 0, 0, 0, 1, 2],
        [0, 2, 1, 0, 0, 1],
        [2, 1, 0, 1, 0, 0]
    ])

    game.render()
    game.make_move(0, 2)

    game.render()

    game.make_move(0, 1)
    game.render()

    game.make_move(5, 2)
    game.render()

    game.make_move(2, 5)
    game.render()

    game.make_move(0, 2)
    game.render()

    game.make_move(5, 1)
    game.render()

    game.make_move(3, 5)
    game.render()

    assert not game.is_valid_move(0, 1)
    print("循环劫测试通过")


def test_banned_moves():
    game = Game(board_size=4)
    game.current_player = 2
    data = """
    x o . x
    . x x .
    . . . .
    . . . .
    """
    game.parse(data)
    assert not game.is_valid_move(0, 2)

    game = Game(board_size=4)
    game.current_player = 2
    data = """
     x o . x
     . x . .
     . . . .
     . . . .
     """
    game.parse(data)
    assert game.is_valid_move(0, 2)

    print("禁止点测试通过")


def test_eat_moves():
    game = Game(board_size=4)
    game.current_player = 1
    data = """
     x o . x
     . x o o
     . . x o
     . . x o
     """
    game.parse(data)
    assert len(game.eat_moves[0, 2]) == 5

    print("打吃点测试通过")


if __name__ == '__main__':
    # 运行测试用例
    test_is_valid_move()
    test_place_stone()
    test_pass_move()
    test_calculate_scores()
    test_place_stone_for_remove()
    test_place_stone_for_remove2()
    test_is_eye()
    test_is_eye_pair()
    test_single_ko_cycle()
    test_ko()
    test_banned_moves()
    test_eat_moves()

File: test/GameTest.cpp
#include "../Game.h"

TEST_CASE("is_valid_move") {
    //边界外落子，无效
    Game game(4);

    CHECK(!game.isValidMove(-1,0));
    CHECK(!game.isValidMove(4,0));
    CHECK(!game.isValidMove(0,-1));
    CHECK(!game.isValidMove(0,4));

    // # 已有棋子的位置落子，应无效
    game.board.board[0][0] = BLACK;
    CHECK(!game.isValidMove(0,0));
}

TEST_CASE("is_eye_pair") {
    Game game(6);
    vector<vector<int> > data = {
        {1, 1, 1, 0, 0, 0},
        {1, 0, 1, 0, 0, 0},
        {1, 1, 0, 1, 0, 0},
        {0, 1, 1, 1, 0, 0},
        {0, 0, 0, 0, 0, 1},
        {0, 0, 0, 0, 0, 1}
    };
    game.board.loadData(data);
    CHECK(game.isEyePair(1,1));
    CHECK(game.isEyePair(2,2));

    data = {
        {1, 1, 1, 0, 0, 0},
        {1, 0, 1, 0, 0, 0},
        {1, 1, 0, 1, 0, 0},
        {0, 1, 1, 1, 1, 1},
        {0, 0, 0, 0, 1, 0},
        {0, 0, 0, 1, 0, 1}
    };
    game.board.loadData(data);
    CHECK(!game.isEyePair(4,5));
    CHECK(!game.isEyePair(5,4));

    data = {
        {0, 1, 0, 1, 0, 1},
        {1, 0, 1, 1, 1, 0},
        {1, 1, 1, 0, 0, 0},
        {0, 0, 1, 1, 1, 1},
        {0, 0, 0, 1, 0, 1},
        {0, 0, 0, 1, 1, 1}
    };
    game.board.loadData(data);
    CHECK(game.isEyePair(0,0));
    CHECK(game.isEyePair(1,1));
}

TEST_CASE("is_eye") {
    Game game(6);
    vector<vector<int> > data = {
        {1, 1, 0, 0, 0, 0},
        {1, 0, 1, 0, 0, 0},
        {1, 1, 1, 0, 0, 0},
        {0, 0, 1, 1, 1, 0},
        {0, 0, 0, 1, 0, 1},
        {0, 0, 0, 1, 1, 1}
    };
    game.board.loadData(data);
    CHECK(game.isEye(1,1));
    CHECK(game.isEye(4,4));

    data = {
        {0, 1, 0, 1, 0, 0},
        {1, 1, 1, 1, 0, 0},
        {1, 1, 1, 0, 0, 0},
        {0, 0, 1, 1, 1, 1},
        {0, 0, 0, 1, 0, 1},
        {0, 0, 0, 1, 1, 1}
    };
    game.board.loadData(data);
    CHECK(game.isEye(0,0));
    CHECK(game.isEye(0,2));
    CHECK(game.isEye(4,4));

    data = {
        {0, 1, 0, 1, 0, 1},
        {1, 0, 0, 1, 1, 0},
        {1, 1, 1, 0, 0, 0},
        {0, 0, 1, 1, 1, 1},
        {0, 0, 0, 1, 0, 1},
        {0, 0, 0, 1, 1, 1}
    };
    game.board.loadData(data);
    CHECK(!game.isEye(0,0));
    CHECK(!game.isEye(0,4));
}

TEST_CASE("calculate_score") {
    Game game(3);
    vector<vector<int> > data = {
        {1, 0, 2},
        {0, 1, 0},
        {2, 0, 1},
    };
    game.board.loadData(data);
    auto pair = game.calculateScore();
    int blackScore = pair.first;
    int whiteScore = pair.second;

    int expected_black_score = 0;
    int expected_white_score = 0;
    for (int i = 0; i < game.boardSize; i++) {
        for (int j = 0; j < game.boardSize; j++) {
            if (game.board.board[i][j] == BLACK)
                expected_black_score++;
            if (game.board.board[i][j] == WHITE)
                expected_white_score++;
        }
    }
    expected_black_score -= game.tieMu;
    expected_white_score += game.tieMu;
    CHECK(expected_black_score == blackScore);
    CHECK(expected_white_score == whiteScore);
}

TEST_CASE("banned_moves") {
    Game game(4);
    vector<vector<int> > data = {
        {1, 2, 0, 1},
        {0, 1, 1, 0},
        {0, 0, 0, 0},
        {0, 0, 0, 0}
    };
    game.board.loadData(data);
    game.currentPlayer = WHITE;
    game.refreshBannedMoves();
    // cout << "banned_moves" << endl;
    // for (auto banned_move: game.bannedMoves)
    //     cout << banned_move.x << " " << banned_move.y << endl;

    CHECK(!game.isValidMove(0,2));

    data = {
        {1, 2, 0, 1},
        {0, 1, 0, 0},
        {0, 0, 0, 0},
        {0, 0, 0, 0}
    };
    game.board.loadData(data);
    game.currentPlayer = WHITE;
    game.refreshBannedMoves();
    CHECK(game.isValidMove(0,2));

    game.boardSize = 5;
    data = {
        {1, 2, 0, 2, 1},
        {0, 1, 1, 1, 1},
        {0, 0, 0, 0, 1},
        {0, 0, 0, 0, 1},
        {0, 0, 0, 0, 1}
    };
    game.board.loadData(data);
    game.currentPlayer = WHITE;
    game.refreshBannedMoves();
    CHECK(!game.isValidMove(0,2));

    data = {
        {1, 2, 0, 2, 1},
        {0, 1, 2, 1, 1},
        {0, 0, 1, 0, 1},
        {0, 0, 0, 0, 1},
        {0, 0, 0, 0, 1}
    };
    game.board.loadData(data);
    game.currentPlayer = WHITE;
    game.refreshBannedMoves();
    CHECK(!game.isValidMove(0,2));

    data = {
        {1, 0, 2, 2, 0},
        {0, 1, 2, 1, 1},
        {0, 0, 1, 0, 1},
        {0, 0, 0, 0, 1},
        {0, 0, 0, 0, 1}
    };
    game.board.loadData(data);
    game.currentPlayer = WHITE;
    game.refreshBannedMoves();
    CHECK(game.isValidMove(0,4));
    CHECK(game.isValidMove(0,1));


    game = Game(9);
    data = {
        {0,2,0,2,1,2,0,2,0},
        {1,2,1,1,1,1,1,1,1},
        {2,2,2,1,2,2,1,2,0},
        {1,2,1,2,1,2,2,2,1},
        {1,2,1,2,0,1,1,2,1},
        {1,1,1,2,1,2,2,2,2},
        {2,2,1,2,1,2,0,2,0},
        {1,0,1,2,1,2,2,1,1},
        {1,2,1,2,2,0,2,1,0}
    };
    game.board.loadData(data);
    game.refreshBannedMoves();
    game.refreshEatMoves();

    CHECK(game.isValidMove(7, 1));


//. x x o x x x x o 
//x x x x x x o x x 
//x . x o x x x x x 
//x x x o o o x x o 
//x x o o o o o o o 
//. x o o . o o . o 
//x x o o o o o o o 
//x o o . o o o o o 
//o o o o x . o o . 

    game = Game(9);
    data = {
        {0,1,1,2,1,1,1,1,2},
        {1,1,1,1,1,1,2,1,1},
        {1,0,1,2,1,1,1,1,1},
        {1,1,1,2,2,2,1,1,2},
        {1,1,2,2,2,2,2,2,2},
        {0,1,2,2,0,2,2,0,2},
        {1,1,2,2,2,2,2,2,2},
        {1,2,2,0,2,2,2,2,2},
        {2,2,2,2,1,0,2,2,0}
    };
    game.board.loadData(data);
    game.currentPlayer = 2;
    game.refreshBannedMoves();
    game.refreshEatMoves();

    auto moves = game.getMoves();
    CHECK(!game.isValidMove(2, 1));
}

TEST_CASE("make_move") {
    Game game(3);
    vector<vector<int> > data = {
        {1, 1, 2},
        {0, 1, 0},
        {2, 0, 1}
    };
    game.board.loadData(data);
    game.refreshEatMoves();


    CHECK(game.eatMoves.size()==1);


    data = {
        {1, 0, 2},
        {0, 1, 0},
        {2, 0, 1}
    };
    game.board.loadData(data);
    game.refreshEatMoves();

    game.makeMove(0, 1);
    game.makeMove(1, 0);
    game.makeMove(1, 2);
    // game.render();

    CHECK(game.board.board[0][2] == NONE_P);

    data = {
        {1, 2, 1},
        {2, 0, 2},
        {1, 2, 1}
    };
    game.board.loadData(data);
    game.currentPlayer = BLACK;
    game.refreshEatMoves();
    game.makeMove(1, 1);
    // game.render();

    CHECK(game.board.board[0][1] == NONE_P);
    CHECK(game.board.board[1][0] == NONE_P);
    CHECK(game.board.board[1][2] == NONE_P);
    CHECK(game.board.board[2][1] == NONE_P);

    game = Game(4);
    data = {
        {1, 2, 0, 1},
        {0, 1, 2, 2},
        {0, 0, 1, 2},
        {0, 0, 1, 2}
    };
    game.board.loadData(data);
    game.refreshEatMoves();

    CHECK(game.eatMoves[Point(0,2)].size()==5);
}


TEST_CASE("test_ko") {
    Game game = Game(4);
    vector<vector<int> > data = {
        {2, 0, 0, 1},
        {0, 2, 1, 0},
        {0, 2, 1, 0},
        {2, 1, 0, 1}
    };
    game.board.loadData(data);
    game.refreshEatMoves();
    game.refreshBannedMoves();
    game.currentPlayer = WHITE;

    game.makeMove(0, 2);
    game.makeMove(0, 1);
    CHECK(!game.isValidMove(0,2));

    game = Game(6);
    data = {
        {2, 0, 0, 1, 0, 0},
        {0, 2, 1, 0, 0, 2},
        {0, 0, 0, 0, 2, 0},
        {0, 0, 0, 0, 1, 2},
        {0, 2, 1, 0, 0, 1},
        {2, 1, 0, 1, 0, 0}
    };
    game.board.loadData(data);
    game.refreshEatMoves();
    game.refreshBannedMoves();
    game.currentPlayer = WHITE;

    game.makeMove(0, 2);
    game.makeMove(0, 1);
    game.makeMove(5, 2);
    game.makeMove(2, 5);
    game.makeMove(0, 2);
    game.makeMove(5, 1);
    game.makeMove(3, 5);

    CHECK(!game.isValidMove(0,1));

    game = Game(4);
    data = {
            {1, 2, 0, 2},
            {0, 1, 2, 2},
            {1, 2, 0, 2},
            {0, 1, 2, 2}
    };
    game.currentPlayer = 1;
    game.board.loadData(data);
    game.refreshEatMoves();
    game.refreshBannedMoves();
    game.recordToHistory();
    game.makeMove(2,2);
    game.render();
    CHECK(!game.isValidMove(2,1));
}

TEST_CASE("eat_move") {
    auto game = Game(4);
    vector<vector<int> > data = {
       {0, 2, 1, 0},
       {2, 2, 1, 0},
       {1, 1, 0, 0},
       {0, 0, 0, 0}
    };
    game.board.loadData(data);
    game.refreshBannedMoves();
    game.refreshEatMoves();
    CHECK(game.isValidMove(0, 0));
}
